<launch>
  <!-- ====== Arguments ====== -->
  <arg name="map_file" default="$(find move_base_navigation)/maps/map.yaml" />
  <arg name="scan_topic" value="/scan"/> 
  <arg name="depth_topic" value="/camera/depth/color/points"/>

  <!-- ====== 센서 실행 (LiDAR & Depth 카메라) ====== -->
  <include file="$(find rplidar_ros)/launch/rplidar_symlink.launch" />

  <include file="$(find realsense2_camera)/launch/rs_camera.launch">
    <arg name="align_depth" value="false" />
    <arg name="enable_color" value="true" />
    <arg name="filters" value="pointcloud"/>
  </include>

  <!-- ====== 맵 서버 실행 ====== -->
  <include file="$(find move_base_navigation)/launch/map_server.launch" />

  <!-- ====== AMCL (측위) 실행 ====== -->
  <include file="$(find move_base_navigation)/launch/amcl.launch">
    <arg name="scan_topic" value="$(arg scan_topic)"/>
  </include>
  
  
  <!-- Move Base 실행 -->
  <node pkg="move_base" type="move_base" name="move_base" output="screen">
  <!-- A*를 전역 플래너로 설정 -->
  <param name="base_global_planner" value="global_planner/GlobalPlanner" />
  
  <!-- TEB를 로컬 플래너로 설정 -->
  <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
  
  <!-- 코스트맵 설정 -->
  <rosparam file="$(find move_base_navigation)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
  <rosparam file="$(find move_base_navigation)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />
  <rosparam file="$(find move_base_navigation)/config/global_costmap_params.yaml" command="load" ns="global_costmap" />
  <rosparam file="$(find move_base_navigation)/config/local_costmap_params.yaml" command="load" ns="local_costmap" />
  
  <!-- Move Base의 기본 매개변수 -->
  <rosparam file="$(find move_base_navigation)/config/move_base_params.yaml" command="load" />
  <!-- D* lite 관련 설정 -->
  <rosparam file="$(find move_base_navigation)/config/global_costmap_params.yaml" command="load" />
  
  <!-- TEB 관련 설정 -->
  <rosparam file="$(find move_base_navigation)/config/teb_local_planner_params.yaml" command="load" />
  </node>
 <!-- costmap_converter 노드 추가 (네비게이션 노드들과 같은 레벨에 추가) -->
<node pkg="costmap_converter" type="standalone_converter" name="costmap_converter_node" output="screen">
  <!-- 사용 가능한 플러그인으로 변경 -->
  <param name="converter_plugin" value="costmap_converter::CostmapToPolygonsDBSConcaveHull"/>
  <param name="obstacle_array_topic" value="/move_base/TebLocalPlannerROS/obstacles"/>
  <!-- 필요시 추가 파라미터 설정 -->
  <param name="/tf/cache_time" value="10.0" />
</node>

  <!-- ====== RViz 실행 ====== -->
  <node pkg="rviz" type="rviz" name="rviz" output="screen" required="true"
        args="-d $(find move_base_navigation)/rviz/visualize_data.rviz"/>
</launch>

